Add AtomicWFQueue implementation and tests

avinashkethineedi opened the pull request:
- [feat: Add AtomicWFQueue implementation](https://github.com/ROCm/rocSHMEM/commit/ec3eb2e2f62dadb2631a3a90200301e91c39029d)
- [test: Add GTest for AtomicWFQueue](https://github.com/ROCm/rocSHMEM/commit/0e32808eba52fed82b24428234095948aea390f5)
- `AtomicWFQueue` will be used for the status flag, g return, and atomic return buffers of the default context.

avinashkethineedi committed:
feat: Add AtomicWFQueue implementation

- Implemented wavefront-safe atomic FIFO queue ensuring first-come, first-serve order
- Added efficient synchronization using atomics
- Enhanced `dequeue` to wait until an element is available

avinashkethineedi committed:
test: Add GTest for AtomicWFQueue

- Implemented unit tests for AtomicWFQueue using GoogleTest framework
- Added tests for `enqueue`, `dequeue`, and edge cases
- Ensured synchronization behavior and correctness under concurrent conditions

edgargabriel commented on src/containers/atomic_wf_queue.hpp:
```diff
@@ -0,0 +1,359 @@
+                             __HIP_MEMORY_SCOPE_AGENT);
+  }
+
+  __device__ void atomic_store(int* address, int val) {
```
NIT: since you used `const int* address` above for the atomic load, shouldn't it also be `const int val` here?

edgargabriel commented on src/containers/atomic_wf_queue.hpp:
```diff
@@ -0,0 +1,359 @@
+/******************************************************************************
+ * Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
NIT: need to determine whether we need to update the year number to 2025

BKP commented on src/containers/atomic_wf_queue_impl.hpp:
```diff
@@ -0,0 +1,164 @@
+     * Ideally, there should always be space available.
+     * If the queue is full, it indicates an unexpected issue.
+     */
+    while (is_full()) {}
```
This should probably be an error state rather than a wait.  It means that something is broken and we should error out.

avinashkethineedi replied:
Yes, I included that to check for any unknown edge cases and to see if the while loop body executes twice. So far, in testing, I haven't seen that while loop running more than once. I'll replace it with the `assert`.

avinashkethineedi committed:
Add assert in `enqueue` and update atomics

- Added an assert in the `enqueue` function to ensure it fails if the queue is full
